contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+which can (optionally) be fractional. The current model is great for supporting
+simple compute resources like CPU or Memory, which are available across all
+kubernetes deployments.
+But there is a problem in representing resources like GPUs, ASICs, local storage
+etc in the form of OIRs. Each of such resource type generally has a rich
+metadata like version, capabilities etc. to describe the resource. A particular
+application pod may perform well only with resources which have a certain
+capability, for example GPUs greater than version 'V'. OIR does not allow such,
+metadata based, selection/filtering of resources.
+The current model requires identity mapping between available resources and requested
+resources. Identity mapping means there must be a one-to-one mapping between the
+resource reference used in the spec to request the resource and resource reference
+used to advertise the resource availablity.
+
+Since 'CPU' and 'Memory' are resources that are available across all kubernetes
+deployments and need no metadata to describe the resource type any further,
+the current user facing API (Pod Specification) remains portable as long as pod
+requests only CPU and Memory. However the current model cannot support
+complex resources like GPUs, ASICs, NICs, local storage, etc.
+To support heterogeneity, portability and management at scale, such resources
+must be represented(advertised) in a form which allows metadata inclusion and a
+metadata based resource selection mechanism.
+
+_GPU Integration Example:_
+  * [Enable "kick the tires" support for Nvidia GPUs in COS](https://github.com/kubernetes/kubernetes/pull/45136)
+  * [Extend experimental support to multiple Nvidia GPUs](https://github.com/kubernetes/kubernetes/pull/42116)
+
+_Kubernetes Meeting Notes On This:_
+  * [Meeting notes](https://docs.google.com/document/d/1Qg42Nmv-QwL4RxicsU2qtZgFKOzANf8fGayw8p3lX6U/edit#)
+  * [Better Abstraction for Compute Resources in Kubernetes](https://docs.google.com/document/d/1666PPUs4Lz56TqKygcy6mXkNazde-vwA7q4e5H92sUc)
+  * [Extensible support for hardware devices in Kubernetes (join kubernetes-dev@googlegroups.com for access)](https://docs.google.com/document/d/1LHeTPx_fWA1PdZkHuALPzYxR0AYXUiiXdo3S0g2VSlo/edit)
+
+## Use Cases
+
+  * I want to have a compute resource type, 'Resource Class',
+    which can be created with meaningful and portable names. This compute
+    resource can hold additional metadata as well, for example:
+    * `nvidia.gpu.high.mem` is the name and metadata is memory greater than 'X' GB.
+    * `fast.nic` is the name and associated metadata is bandwidth greater than
+      'B' gbps.
+  * If I request a resource `nvidia.gpu.high.mem` for my pod, any 'nvidia-gpu'
+    type device which has memory greater than or equal to 'X' GB, should be able
+    to satisfy this request, independent of other device capabilities such as
+    'version' or 'nvlink locality' etc.
```

Comment:
"high" is subjective, though, right? Anything that is subjective is guaranteed to be non-portable across Kubernetes deployments. One k8s deployment might consider 2G to be high. Another might consider 8G to be high. This is the problem with using a qualitative string like "nvidia.gpu.high.mem" to describe a quantitative concept. Instead, I'd recommend segregating the quantitative concepts from the qualitative things.

For instance, consider making the resource classes strictly quantitative:
* `vgpu`
* `vgpu_display_head`
* `gpu_memory_bytes`

etc. And then have a separate set of qualitative strings/metadata that would allow the provider of those quantitative resources to be further described:

* `gpu.opencl.1_0`
* `gpu.opencl.1_1`
* `gpu.opencl.1_2`
* `gpu.opencl.2_0`
* etc..

That way the user can request 2 vGPUs and 4 display heads from something that provides OpenCL version 2.0 support. The separation of quantitative things and qualitative things is clean.

Reply:
Thanks for this nice example to highlight the qualitative vs quantitative aspects.

Reply:
>"high" is subjective, though, right? Anything that is subjective is guaranteed to be non-portable across Kubernetes deployments. One k8s deployment might consider 2G to be high. Another might consider 8G to be high. 

@jaypipes  `nvidia.gpu.high.mem` is just a reference to the resource class object, ex:
```
kind: ResourceClass
metadata:
  name: nvidia.high.mem
spec:
  resourceSelector:
    - matchExpressions:
        - key: "Kind"
          operator: "In"
          values:
            - "nvidia-gpu"
        - key: "memory"
          operator: "GtEq"
          values:
            - "30G"
```

"high" is not a selection parameter rather properties are, and properties are quantitative. Using above res class example, this is a resource class of GPUs which have memory greater than 30 GB. As long as resource class objects are same across clusters, there is no problem in app/pod portability across these clusters. WDYT?

Reply:
@vikaschoudhary16 My point was that the resource class object is mixing quantitative and qualitative things. It's mixing quantitative things like amount of memory with qualitative things like "Kind: nvidia". By doing so, I suspect you'll end up in a situation where you have a whole bunch of resource classes that only differ slightly on some minor qualitative things (or minor quantitative things). For example, you may end up with separate resource class objects for various permutations of vendor, pGPU type, OpenCL/CUDA software version support, display heads, etc.

Is the idea to have lots of these resource class objects exposed to the user that they must search through to determine the exact combination of qualitative and quantitative things that the resource class object represents? Or is the idea to allow the user to specify the more granular resource quantitative and qualitative parts of the request in-line in the ContainerSpec's ResourceRequirements.requests object specification?

I'm trying to determine whether these resource classes will end up being a special snowflake way of requesting *some* kinds of resources instead of using the existing ResourceRequirements.resource/limits ways of specifying "normal" resource amounts.

Best,
-jay

Reply:
@jaypipes My idea is that Resource classes will be a way to offer different flavors of devices. Just to give an analogy like ec2 instance types. End user will just request resource class in the pod. resource classes will be RBAC enabled. End user will not be able to modify/update resource class. 

> I'm trying to determine whether these resource classes will end up being a special snowflake way of requesting some kinds of resources instead of using the existing ResourceRequirements.resource/limits ways of specifying "normal" resource amounts.

resource class will be requested using the existing ResourceRequirements.resource/limits way and can be seen as an abstracted resource.

Thoughts?

Reply:
Getting too selective on non-fungible resources can be a dangerous game vs. channeling a lowest common denominator.  

For a decade on grid systems we allow arbitrary matching on any attribute via an expressive configuration language, and it was eventually highly abused by its users to hoard the prized resources.